Vim.Compat: Add several compatible functions

lambdalisue opened the pull request:
https://github.com/vim-jp/vital.vim/issues/343

ujihisa commented on doc/vital-vim-compat.txt:
```diff
@@ -34,6 +35,34 @@ has_version({version})			*Vital.Vim.Compat.has_version()*
+redraw([{bang}])			*Vital.Vim.Compat.redraw()*
+	Compatibility function of |redraw| which has an issue on MacVim (GUI).
+	It calls |sleep| 1m just after |redraw| when called in MacVim (GUI),
+	otherwise it just calls |redraw|.
```
`s/call/execute/g` because `redraw` is not a function but a command.

lambdalisue commented:
MacVim's redraw/redrawstatus issue will be fixed https://github.com/macvim-dev/macvim/issues/174#issuecomment-164827421

I'm not sure if `Vim.Compat.redraw()` should exist or not now.

rhysd commented:
Please read documentation.

> *Vital.Vim.Compat* provides vim compatibility wrapper functions for different
> versions/patchlevels of Vim.
>
> This module is not for multiple OS compatibilities but for versions of Vim
> itself.

`Vim.Compat` aims to absorb the difference among versions/patchlevels of Vim.  I think `redraw()` should be removed.

lambdalisue commented:
Yes I first thought so as well but the word version can indicate "MacVim's specific version" and philosophy of the module stands for compatibility so I don't think the function against to the purpose of the module.

But as I mentioned, the feature is fixed today and (I think) people usually use HEAD version so, in this point of view, I'm not sure if the function should exist or force users to use a latest version of MacVim.

lambdalisue commented:
Additional note:
redraw compat comes from a bug of (Mac)Vim, not because of OS difference. That why I think it should be in Vim.Compat if it is required.

rhysd commented:
I agree that this is a problem of the term 'Vim compatibility'.

If the bug is included in 'compatibility' which this module must cover,  too many bugs should be covered by this module.  As you know, many bugs are reported and fixed by Vim developers.

I think this module should cover the API compatibility such as `glob()` and should not cover buggy behavior.  (Sure, I agree that it is useful to wrap `redraw()`.)

lambdalisue commented:
> If the bug is included in 'compatibility' which this module must cover, too many bugs should be covered by this module. As you know, many bugs are reported and fixed by Vim developers.

Completely true. I changed my mind ;-)

lambdalisue commented:
I just remove a workaround function for a `redraw` command while:

1. As @rhysd said, the module should **ONLY** stand for covering compatibility of difference of Vim's version and not for workarounds for known bugs.
2. `redraw` issue has been fixed from [this commit](https://github.com/macvim-dev/macvim/commit/e7ca3fe78949b61f1a4962075b638ceb46645195) on macvim-dev/macvim repository and the command `homebrew install --HEAD macvim`, which most of MacVim users use recently, installs the latest version.

Probably developers need to add `Vim.Compat.Workarounds.redraw()`, `Vim.MacVim.redraw()`, or whatever to add this function **if necessary**. For now, I don't think the workaround function is required anymore ;-)

crazymaster commented:
>`https://github.com/vim-jp/vim/commit/8399b184df06f80ca030b505920dd3e97be72f20`

These links is not found.

lambdalisue commented:
> These links is not found.

Hum... OK I'll remove these thanks

ujihisa commented:
:+1: :sushi:

rhysd commented:
@lambdalisue 

Thank you for understanding.  Looks good to me.  I think this PR is ready to merge after resolving conflict with `master` branch.

ujihisa commented:
looks like it's mergeable

rhysd commented:
@ujihisa 

Hmm,  at least in my browser it says 'This branch is out-of-date with the base branch'.

![tmp](https://cloud.githubusercontent.com/assets/823277/11888428/d553c180-a580-11e5-94a5-8833b4278d81.png)

ujihisa commented:
!?

![](https://cache.gyazo.com/b6a20d0b2ee3b6e9061f52a145fffdba.png)

rhysd commented:
lol GitHub says @ujihisa should merge this.

ujihisa commented:
:shipit: merge力 :shipit:

rhysd commented:
:muscle::shipit:

lambdalisue commented:
What lol. Anyway thanks ;-)

rhysd commented:
Nice work, @lambdalisue !
